crypto/bn/asm/loongarch-mont.pl

```diff
@@ -0,0 +1,333 @@
+#! /usr/bin/env perl
+# Copyright 2010-2018 The OpenSSL Project Authors. All Rights Reserved.
+#
+ 
+# Licensed under the Apache License 2.0 (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+#
+# ====================================================================
+# Written by Song Ding <songding@loongson.cn> for the OpenSSL
+# project. The module is, however, dual licensed under OpenSSL and
+# CRYPTOGAMS licenses depending on where you obtain it. For further
+# details see http://www.openssl.org/~appro/cryptogams/.
+# ====================================================================
+
+# December 2021
+#
+# This module doesn't present direct interest for OpenSSL, because it
+# doesn't provide better performance for longer keys, at least not on
+# in-order-execution cores. While 512-bit RSA sign operations can be
+# 65.7% faster in 64-bit mode, 1024-bit ones are 24.6% faster, and
+# 2048-bit ones are only 10.4% faster, the code improve rsa512 *verify*
+# benchmark by 20.4%, rsa1024 one - by 10.9%:-)
+# All comparisons are against bn_mul_mont-free assembler.
+
+######################################################################
+# Here is register layout for LOONGARCH ABIs.
+# The return value is placed in $v0($a0).
+
+($zero,$ra,$tp,$sp,$fp)=map("\$r$_",(0..3,22));
+($a0,$a1,$a2,$a3,$a4,$a5,$a6,$a7)=map("\$r$_",(4..11));
+($t0,$t1,$t2,$t3,$t4,$t5,$t6,$t7,$t8)=map("\$r$_",(12..20));
+($s0,$s1,$s2,$s3,$s4,$s5,$s6,$s7,$s8)=map("\$r$_",(23..31));
+
+
+$PTR_ADD="addi.d";
+$REG_S="st.d";
+$REG_L="ld.d";
+$SZREG=8;
+
+######################################################################
+
+while (($output=shift) && ($output!~/\w[\w\-]*\.\w+$/)) {}
+open STDOUT,">$output";
+
+$LD="ld.d";
+$ST="st.d";
+$MULD="mul.d";
+$MULHD="mulh.du";
+$ADDU="add.d";
+$SUBU="sub.d";
```

Comment:
Please don't pretend you're MIPS. The `u` suffix doesn't mean the same in LoongArch assembly as in MIPS, it's misleading.